core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/InternalMultiFieldStats.java

```diff
@@ -0,0 +1,254 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.metrics.stats.multifield;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentBuilderString;
+import org.elasticsearch.search.aggregations.AggregationStreams;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.metrics.InternalMetricsAggregation;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Computes distribution statistics over multiple fields
+ */
+public class InternalMultiFieldStats extends InternalMetricsAggregation implements MultiFieldStats {
+
+    public final static Type TYPE = new Type("multifield_stats");
+    public final static AggregationStreams.Stream STREAM = new AggregationStreams.Stream() {
+        @Override
+        public InternalMultiFieldStats readResult(StreamInput in) throws IOException {
+            InternalMultiFieldStats result = new InternalMultiFieldStats();
+            result.readFrom(in);
+            return result;
+        }
+    };
+
+    public static void registerStreams() {
+        AggregationStreams.registerStream(STREAM, TYPE.stream());
+    }
+
+    /** per shard stats needed to compute stats */
+    protected MultiFieldStatsResults multiFieldStatsResults;
+
+    protected InternalMultiFieldStats() {
+    }
+
+    protected InternalMultiFieldStats(String name, long count, MultiFieldStatsResults multiFieldStatsResults,
+                                      List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) {
+        super(name, pipelineAggregators, metaData);
+        assert count >= 0;
+        this.multiFieldStatsResults = multiFieldStatsResults;
+    }
+
+    @Override
+    public Type type() {
+        return TYPE;
+    }
+
+    @Override
+    public long getDocCount() {
+        return multiFieldStatsResults.docCount;
+    }
+
+    @Override
+    public long getFieldCount(String field) {
+        if (multiFieldStatsResults == null || multiFieldStatsResults.means.containsKey(field) == false) {
+            return 0;
+        }
+        return multiFieldStatsResults.counts.get(field);
+    }
+
+    @Override
+    public double getMean(String field) {
+        if (multiFieldStatsResults == null || multiFieldStatsResults.means.containsKey(field) == false) {
+            return Double.NaN;
```

Comment:
General thought for these various `Double.NaN`'s .. should we be returning `null` instead (and subsequently returning a `Double` instead of `double`)?  E.g. because some fields may actually have `NaN`'s as real values, and lacking a field isn't quite the same as not-a-number?

Not sure, probably needs several more opinions :)